### PR TITLE
Unsupervised Intrusion Detection via Swarm Learning

### PR DESCRIPTION
**Pull request by mahi4905**
This PR introduces a 3-node Swarm Learning pipeline for binary network intrusion detection on the NSL-KDD dataset, using unsupervised Deep Embedded Clustering rather than a directly supervised classifier.
 
* **Binary Filter (Deep K-Means Autoencoder):** Trains an unsupervised autoencoder + K-Means clustering head across 3 simulated Swarm nodes, learning to separate `normal` from `attack` traffic without ever training on the label directly.
* **Non-IID Node Splits:** Each node trains on a deliberately skewed partition of `KDDTrain+.txt` — one normal-dominant (65/35), one attack-dominant (65/35), one balanced (50/50) — to exercise Swarm Learning's merge behavior under realistic, non-identical data distributions, without nodes ever sharing raw traffic logs.
* **Unsupervised NMI Purity:** Because this relies on unsupervised clustering, cluster quality is evaluated using Normalized Mutual Information (NMI) against the true labels. Across the 3 nodes' final epoch, NMI ranged from **0.3238 to 0.3552**, indicating the learned clusters carry a real, non-trivial relationship to the true normal/attack split despite never training on it directly.
By running this through `train_binary.py` with `SwarmCallback` synchronization, all 3 nodes complete the full 50-epoch training run cleanly (no inter-node sync failures), each producing an isolated final model and cluster map evaluable against the held-out `KDDTest+.txt` set via `evaluate_binary.py`.

### CODE REVIEW

**Comment by mahi4905** [comment]
sn1 logs
mahi@instance-20260415-181739:~/swarm-learning/workspace/nsl-kdd-v8$ docker logs sn1
######################################################################
##                    HPE SWARM LEARNING SN NODE                    ##
######################################################################
## © Copyright 2019-2023 Hewlett Packard Enterprise Development LP  ##
######################################################################
2026-06-25 05:37:59,006 : swarm.blCnt : INFO : Setting up blockchain layer for the swarm node: START
2026-06-25 05:38:00,456 : swarm.blCnt : INFO : Creating Autopass License Provider
2026-06-25 05:38:01,344 : swarm.blCnt : INFO : Creating license server
2026-06-25 05:38:01,344 : swarm.blCnt : INFO : Setting license servers
2026-06-25 05:38:01,355 : swarm.blCnt : INFO : Acquiring floating license 1100000380:1
2026-06-25 05:38:35,402 : swarm.SN : INFO : SMLETHNode: Starting GETH ... 
2026-06-25 05:38:45,449 : swarm.SN : WARNING : SMLETHNode: Enode list is empty: Node is standalone
2026-06-25 05:44:27,747 : swarm.SN : INFO : SMLETHNode: Started I-am-Alive thread
2026-06-25 05:44:27,747 : swarm.blCnt : INFO : Setting up blockchain layer for the swarm node: FINISHED
2026-06-25 05:44:32,382 : swarm.blCnt : INFO : Starting SWARM-API-SERVER on port: 30304
2026-06-25 05:58:17,960 : swarm.blCnt : INFO : Received termination signal. Stopping the SWARM-API-SERVER
2026-06-25 05:58:17,961 : swarm.blCnt : INFO : Stopping blockchain
2026-06-25 05:58:18,200 : swarm.blCnt : INFO : Stopped blockchain, return code=0
2026-06-25 05:58:18,201 : swarm.blCnt : WARNING : Releasing license

**Comment by mahi4905** [comment]
log files
[sn1.log](https://github.com/user-attachments/files/29412786/sn1.log)
[ml1.log](https://github.com/user-attachments/files/29412792/ml1.log)
[sl1.log](https://github.com/user-attachments/files/29412795/sl1.log)
[ml2.log](https://github.com/user-attachments/files/29412796/ml2.log)
[sl2.log](https://github.com/user-attachments/files/29412797/sl2.log)
[ml3.log](https://github.com/user-attachments/files/29412798/ml3.log)
[sl3.log](https://github.com/user-attachments/files/29412801/sl3.log)
[result.log](https://github.com/user-attachments/files/29412803/result.log)

**Comment by mahi4905** [comment]
Working(Youtube (link-only view))
[Youtube Link](https://youtu.be/o4iaw9Ee7q8)

File: examples/unsupervised-k-means/data-prep/prepare_data.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import os
```
Comment (RadhakrishnaJ) [inline examples/unsupervised-k-means/data-prep/prepare_data.py:33]: Get the copy right like in other examples...  use year as 2026


Remember to cross check in other files also...

File: examples/unsupervised-k-means/data-prep/prepare_data.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    .drop(columns="difficulty")
+    .reset_index(drop=True)
+)
+
+print(f"Total rows: {len(df)}")
+print(df["label"].value_counts())
+
+normal_df = df.query("label == 'normal'")
+attack_df = df.query("label == 'attack'")
+
```
Comment (RadhakrishnaJ) [inline examples/unsupervised-k-means/data-prep/prepare_data.py:86]: Add some comments about data distribution strategy. Also, throughout all files add comments about functionality.

File: examples/unsupervised-k-means/readme.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
+# NSL-KDD Binary Intrusion Detection
+
```
Comment (RadhakrishnaJ) [inline examples/unsupervised-k-means/readme.md:2]: Include the citation of actual Deep K means paper and also add in comments section of code

Comment (mahi4905) [reply examples/unsupervised-k-means/readme.md:2]: I have addressed all three comments in the latest commit — added the (C)Copyright 2021-2026 HPE Apache 2.0 header to all Python files, added inline comments explaining the non-IID data distribution strategy and functionality throughout all files, and added the DEC paper citation (Xie et al., ICML 2016) to both README.md and train_binary.py

File: examples/unsupervised-k-means/data-prep/prepare_data.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
+############################################################################
+## (C)Copyright 2021-2026 Hewlett Packard Enterprise Development LP
```
Comment (suresh-ls) [inline examples/unsupervised-k-means/data-prep/prepare_data.py]: Since this is a completely new file and we are just adding it now,   the year can be just "2026" ( no range needed )   The same comment in other files where ever this is applicable.

We put 'Start year-End year' when the work was started in the 'Start year' and the same file is updated now in a particular year.

File: examples/unsupervised-k-means/model/train_binary.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+#   SwarmCallback must wrap a SINGLE continuous model.fit() call.
+#   Calling fit() multiple times re-triggers on_train_end() each
+#   time, signaling the SL container that training is complete
+#   prematurely and causing a broken-pipe error on the next sync.
+#   The periodic target-distribution refresh required by DEC is
+#   handled inside DeepKMeansCallback using a tf.data.Dataset
+#   generator, allowing P to update each epoch within one fit().
+#
+# Reference:
+#   Xie, J., Girshick, R., & Farhadi, A. (2016).
```
Comment (suresh-ls) [inline examples/unsupervised-k-means/model/train_binary.py:48]: Did we not use this as the main reference paper ?  Why are we not quoting this here ?

Deep k-Means: Jointly clustering with k-Means and learning representations
[Maziar Moradi Fard] [Thibaut Thonet], [Eric Gaussier]
https://arxiv.org/abs/1806.10069) [cs.LG]

**Comment by RadhakrishnaJ** [comment]
New comments - 1. Remove your host ip from instructions... use ${HOST_1_IP} like this, refer other examples. 
2. In Copyright, mention only 2026 because these files are new.

**Comment by mahi4905** [comment]
I've addressed the requested changes in the latest commit Thank you!

**Review by RadhakrishnaJ — APPROVED** [review]
Approve - The proposed updates look good, and the review feedback has been incorporated. I have not personally verified the execution by running the provided instructions; however, based on the attached logs, the changes appear to execute as intended.

File: examples/unsupervised-k-means/model/train_binary.py (hunk 5 above)
Comment (suresh-ls) [reply examples/unsupervised-k-means/model/train_binary.py:48]: Did we not use the above reference paper/work ?  I thought we used this and converted the code from TF1.1 to TF2.0 ?

If so, can we mention this work in  both the places where references are used .

**Comment by suresh-ls** [comment]
I have a important question on which paper we have referred..  Please look into the previous comment, and address that.

**Comment by mahi4905** [comment]
I've added the Deep k-Means paper and the reference implementation repository to the README in the latest commit

**Review by suresh-ls — APPROVED** [review]
Approved - The proposed updates look good and the review feedback has been incorporated.
